msos/WaitChainTraversal.cs

```diff
@@ -0,0 +1,216 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+using static msos.NativeStructs;
+using static msos.NativeMethods;
+
+namespace msos
+{
+    /// <summary>
+    /// Exract blocking objects information using WCT api:
+    /// doc: https://msdn.microsoft.com/en-us/library/windows/desktop/ms681622(v=vs.85).aspx
+    /// </summary>
+    class WaitChainTraversal
+    {
+        public const int WCT_MAX_NODE_COUNT = 16;
+        public const uint WCTP_GETINFO_ALL_FLAGS = 7;
+        public const int WCT_OBJNAME_LENGTH = 128;
+
+        /// <summary>
+        /// Gets WCT information using WCT api by the given threadId (OS threadId)
+        /// </summary>
+        /// <param name="threadId"> ThreadWCTInfo  object</param>
+        /// <returns>Thread id of thread target</returns>
+        public ThreadWCTInfo GetBlockingObjects(uint threadId)
+        {
+            ThreadWCTInfo result = null;
+
+            var g_WctIntPtr = OpenThreadWaitChainSession((int)WCT_SESSION_OPEN_FLAGS.WCT_SYNC_OPEN_FLAG, 0);
+
+
+            WAITCHAIN_NODE_INFO[] NodeInfoArray = new WAITCHAIN_NODE_INFO[WCT_MAX_NODE_COUNT];
+
+
+            int isCycle = 0;
+            int Count = WCT_MAX_NODE_COUNT;
+
+            // Make a synchronous WCT call to retrieve the wait chain.
+            bool waitChainResult = GetThreadWaitChain(g_WctIntPtr,
+                                    IntPtr.Zero,
+                                    WCTP_GETINFO_ALL_FLAGS,
+                                    threadId, ref Count, NodeInfoArray, out isCycle);
+
+            CheckCount(ref Count);
+
+            if (waitChainResult)
+            {
+                result = HandleGetThreadWaitChainRsult(threadId, Count, NodeInfoArray, isCycle);
+            }
+            else
+            {
+                HandleWctRequestError();
+            }
+
+            //Finaly ...
+            CloseThreadWaitChainSession(g_WctIntPtr);
+
+            return result;
+        }
+
+        private ThreadWCTInfo HandleGetThreadWaitChainRsult(uint threadId, int Count, WAITCHAIN_NODE_INFO[] NodeInfoArray, int isCycle)
+        {
+            ThreadWCTInfo result = new ThreadWCTInfo(isCycle == 1, threadId);
+            WAITCHAIN_NODE_INFO[] info = new WAITCHAIN_NODE_INFO[Count];
+            Array.Copy(NodeInfoArray, info, Count);
+
+            result.SetInfo(info);
+
+            return result;
+        }
+
+        private void HandleWctRequestError()
+        {
+            var lastErrorCode = Marshal.GetLastWin32Error();
+
+            if (lastErrorCode == (uint)SYSTEM_ERROR_CODES.ERROR_IO_PENDING)
+            {
+                //TODO: Follow this doc: https://msdn.microsoft.com/en-us/library/windows/desktop/ms681421(v=vs.85).aspx
```

Comment:
Are these comments still relevant? What's the handling that needs to happen here?